Conform appbar buttons to Material

xster opened the pull request:
#5201
- Extend left to edge, right to 4dp
- Make leading button square and 56dp
- Keep title at 72dp on Android according to Material
- Take injected actions buttons and expand to recommended 48dp width (also lifted from other Google apps) if it's an IconButton or PopupMenuButton
- Test

xster committed:
Conform leading appbar button to Material

- Extend to edge
- Make square and 56dp
- Test

xster committed:
Expand action buttons and test

xster committed:
Add test for not messing with non-IconButton

xster commented:
![flutter_01](https://cloud.githubusercontent.com/assets/156888/23089984/b3efd9e4-f547-11e6-926f-03118babfe1c.png)

abarth commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -16,6 +16,7 @@ import 'icon_theme.dart';
 import 'icon_theme_data.dart';
 import 'icons.dart';
 import 'material.dart';
+import 'popup_menu.dart';
```
We haven't been importing libraries just to make dartdoc links work.  Instead, our plan is to teach dartdoc how to link to classes you haven't imported.

xster replied:
It's for the `action is PopupMenuButton` but let's continue on that thread

xster replied:
Removed this

abarth commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -388,19 +392,29 @@ class AppBarState extends State<AppBar> {
     if (config.actions != null && config.actions.isNotEmpty) {
+      // Expand to at least 48dp if IconButton or PopupMenuButton.
+      List<Widget> sizedActions = config.actions.map<Widget>((Widget action) {
+        return action is IconButton || action is PopupMenuButton ?
```
Runtime type checks like these break composibility.  For example, if `action` is a widget that builds an `IconButton`, it will work differently than a straight `IconButton`.  Similarly, if you copy and paste the code for `IconButton` into your own codebase and change its name to `IconButton2`, then it will behave differently.

Instead, we should just apply this constraint to all the children, regardless of what type there are.  That is to say, AppBar should have the opinion that the actions have a min width of 48.0.

Also, we tend not to use `map` because it's less efficient than just writing out the loop.

xster replied:
Ah, got it. We can expand everything. Though if action is a container that has a constraint that has a gesture detector, we might not increase the actual touch target. I guess we can just document that. 

> Also, we tend not to use map because it's less efficient than just writing out the loop.

That's crazy lol. Should we file a bug? Feels like map should be more parallelizable.

abarth replied:
> That's crazy lol. Should we file a bug? Feels like map should be more parallelizable.

Parallelizable in what sense?  The map evaluates the given function over each element of the list in order sequentially.

xster replied:
Ah, didn't know sequentiality was guaranteed. 

Moved the whole expansion thing to icon button.

abarth commented:
Should IconButtons default to being 48x48 all the time?  I wonder if the default padding is wrong.

xster commented:
Ya, we should definitely reduce friction to have the 48dp min touch target. Filed #8264

xster committed:
Remove widget type handling

xster committed:
Remove popup menu import

xster committed:
Move 48dp size expansion to icon button

xster commented:
- Renamed IconButton.size to .iconSize
- Moved 48dp size expansion from app bar to icon button
- IconButton default constraints to 48. Can still stretch to infinity but can't be smaller than 48.
- Ink splash matches touch target size (plus a bit more)
- Tests

abarth commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -339,7 +339,7 @@ class AppBarState extends State<AppBar> {
         leading = new IconButton(
           icon: new Icon(Icons.menu),
-          alignment: FractionalOffset.centerLeft,
+          alignment: FractionalOffset.center,
```
This is the default value.  We can remove it.

xster replied:
Done

abarth commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -358,7 +358,7 @@ class AppBarState extends State<AppBar> {
           leading = new IconButton(
             icon: new Icon(backIcon),
-            alignment: FractionalOffset.centerLeft,
+            alignment: FractionalOffset.center,
```
This is the default value.  We can remove it.

xster replied:
Done

abarth commented on packages/flutter/lib/src/material/icon_button.dart:
```diff
@@ -136,29 +138,30 @@ class IconButton extends StatelessWidget {
+      constraints: new BoxConstraints(minWidth: 48.0, minHeight: 48.0),
+      child: new Padding(
+        padding: padding,
         child: new ConstrainedBox(
```
I think you can simply this to just be a `SizedBox.fromSize(size: iconSize, ...`

xster committed:
Merge branch 'master' of github.com:flutter/flutter into appbar-1

abarth commented on packages/flutter/lib/src/material/icon_button.dart:
```diff
@@ -136,29 +138,30 @@ class IconButton extends StatelessWidget {
           )
         )
-      )
+      ),
```
nit: the current style guide is to add trailing commas to all of these lines.

xster replied:
Done

abarth commented on packages/flutter/lib/src/material/icon_button.dart:
```diff
@@ -136,29 +138,30 @@ class IconButton extends StatelessWidget {
-        maxHeight: size,
+
+    Widget result = new ConstrainedBox(
+      constraints: new BoxConstraints(minWidth: 48.0, minHeight: 48.0),
```
I'd extract the 48.0 as a private constant in this file.

xster replied:
Done

abarth commented on packages/flutter/lib/src/material/icon_button.dart:
```diff
@@ -168,7 +171,8 @@ class IconButton extends StatelessWidget {
       child: result,
-      radius: math.max(size, Material.defaultSplashRadius),
+      radius: math.max(48.0, iconSize + math.min(padding.horizontal, padding.vertical))
+          * 0.55, // x 0.5 for diameter -> radius and + 10% overflow so it doesn't look too square.
```
I'm unsure about this change.  Is there anything in the spec that says what these values should be?  I would expect in the default (48x48) case, that Material.defaultSplashRadius is the proper radius.  I think we got that by measuring something in the spec, but I'm not certain.

Your calculation looks reasonable for the non-default case.  I just wonder if we should use Material.defaultSplashRadius rather than using 48.0 * 0.55 as the left parameter to the max.

xster replied:
Material.defaultSplashRadius is a static number right? I didn't find anything specific in the material guidelines. Though looking at images in the website and other material apps, a 70px splash on a 48px button with a 24px icon looks too big. 

Let's look at it together tomorrow.

abarth replied:
> Material.defaultSplashRadius is a static number right?

Yes.

> I didn't find anything specific in the material guidelines.

Ok, I think I found my source.  Check out this video:

https://storage.googleapis.com/material-design/publish/material_v_10/assets/0B14F_FSUCc01MmtYUVI0VE85Y00/02_SearchRipple-v2.webm

Here's a screen capture when the "inner" splash on the right reaches its largest radius:

![inkreaction](https://cloud.githubusercontent.com/assets/112007/23196205/0f831758-f86d-11e6-94f0-459beeefe48b.png)

In the video, the radius of that splash (not the largest pink circle but the one inside that one (not the darkest pink ring), is 56.0 pixels.  However, the video doesn't appear to be 1-to-1 because the toolbar is 90.0 pixels.  We know from another part of the spec that the toolbar is actually 56.0 logical pixels, so that puts the radius of the splash at 56/90*56 = 34.8 logical pixels (which I think we rounded to 35.0).

xster replied:
Ya, you're totally right. The contacts app has a 34.33 radius on a presumably 48 icon. Will keep the 35 and also kept the 40% overflow for bigger sizes.

abarth commented:
![LGTM](http://i3.cpcache.com/product/1351564863/lgtm_bumper_stickers.jpg?width=200&height=200)

... but Travis is mad

xster committed:
Add touch handling test

xster committed:
Review notes

xster committed:
Merge branch 'master' of github.com:flutter/flutter into appbar-1

xster committed:
bigger ink splash

Hixie commented:
I have no idea why but this improved performance on iOS complex_layout quite substantially.

xster commented:
Could it be a coincidence? I don't have theory of why it might be. Unless there's something with the implementation of LimitedBox.
